cmd/certs.go

```diff
@@ -20,51 +20,293 @@ import (
+	if !isFile(keyFile) && globalCLIContext.AutoTLS != "" {
+		if err := generatePrivateKey(keyFile); err != nil {
+			return nil, nil, err
+		}
+		if isFile(certFile) { // delete any existing cert
+			if err := os.Remove(certFile); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	// If there is no private key and no certificate we don't use TLS.
+	// In case of auto-tls we've created the private key just before and
+	// therefore there is at least a private key.
+	if !isFile(certFile) && !isFile(keyFile) {
+		return nil, nil, nil
+	}
+
+	// Now load the private key which was either generated by the user
+	// or, in case of auto-tls, by us previously.
+	privateKey, publicKey, privatePEMKey, err := loadPrivateKey(keyFile)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If there is no certificate and the auto-tls argument is present
+	// generate a new certificate. The auto-tls arguments are the
+	// SANs (as comma-separated list).
+	if !isFile(certFile) && globalCLIContext.AutoTLS != "" {
+		if err = generateCertificate(certFile, privateKey, publicKey, strings.Split(globalCLIContext.AutoTLS, ",")); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// Load and decode the certificate which is expected to be a
+	// valid PEM-encoded ANSI.1 string.
+	rawCert, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, nil, uiErrTLSNoCertificate(err)
+	}
+	certificates, err := decodeCertificates(rawCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If the certificate has no intermediate or root CA certificates as parents
+	// we assume it's a "self-signed" certificate. In case of auto-tls we check
+	// wehether it's expired and if so renew it.
+	if len(certificates) == 1 && globalCLIContext.AutoTLS != "" {
+		x509Cert := certificates[0]
+		if now := time.Now().UTC(); now.Before(x509Cert.NotBefore.UTC()) || now.After(x509Cert.NotAfter.UTC()) {
+			if err = generateCertificate(certFile, privateKey, publicKey, strings.Split(globalCLIContext.AutoTLS, ",")); err != nil {
+				return nil, nil, err
+			}
+			rawCert, err = ioutil.ReadFile(certFile)
+			if err != nil {
+				return nil, nil, uiErrTLSNoCertificate(err)
+			}
+			certificates, err = decodeCertificates(rawCert)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	// Create a TLS certificate from the PEM certificate and the PEM private key.
+	// If the private key does not match the certficate this will fail. That can happen
+	// if e.g. the certificate was generated wrongly by external tools.
+	certificate, err := tls.X509KeyPair(bytes.TrimSpace(rawCert), privatePEMKey)
+	if err != nil {
+		return nil, nil, uiErrTLSCertificateCreationFailed(err)
+	}
+	return &certificate, certificates, nil
+}
+
+// generatePrivateKey generates a new random private key and
+// stores it at keyFile. If the env. variable `MINIO_CERT_PASSWD`
+// is present it encrypts the private key before writing it to
+// the file.
+func generatePrivateKey(keyFile string) error {
```

Comment:
shall we move this generate code into `pkg/certs` ?

Reply:
It is not really independent since it looks up the `TLSPrivateKeyPassword` environment variable. However, we could if this considered okay...

Reply:
Yeah why not, packages can support envs independently.